Render Architecture Overview & Pipelined Rendering Book Pages (#2469)
bevyengine/bevy-website: content/learn/book/the-renderer/render-pipelines.md

This is a linear process, and can be restrictive: the simulation can't progress until the render is finished, and the render can't start until the simulation is complete.
While this does still work for less intensive or older games, the abilities of modern computer hardware have outgrown this approach.
Both simulations and rendering logic have become more complicated in recent years, meaning that it isn't performant for one process to wait for the other to finish.

**Pipeline Rendering** is an alternative approach that sees the main simulation logic decoupled from the render logic.
Instead of waiting on each other, the simulation logic is run on one CPU thread while the render logic is copied to a separate thread before being passed to the GPU to be rendered.
This allows the simulation to execute concurrently with the render, meaning less "idle" time for both processes.

Bevy implements a version of pipelined rendering using two separate `SubApp`s to track the compute and render logic separately.
Both `SubApp`s abide by ECS principles and have their own `World`s, schedules, and systems to control how their respective logic is executed.

> Specifically, the Compute `SubApp` (where our Main `World` lives) calculates the simulation logic.

Reviewer: ```suggestion
Specifically, the Main `SubApp` (where our Main `World` lives) calculates the simulation logic.
```

> Then, the freshly computed info in the Compute `SubApp` is copied and extracted into a Render `SubApp`.

Reviewer: ```suggestion
Then, the freshly computed info in the Main `SubApp` is extracted into the Render `SubApp`.
```

+++
title = "Pipelined Rendering"
insert_anchor_links = "right"
[extra]
weight = 2
+++

Traditional rendering methods usually rely on the "simulation" logic (game state, path-finding, physics, etc.) being computed and then "rendered" to the screen after.
This is a linear process, and can be restrictive: the simulation can't progress until the render is finished, and the render can't start until the simulation is complete.
While this does still work for less intensive or older games, the abilities of modern computer hardware have outgrown this approach.
Both simulations and rendering logic have become more complicated in recent years, meaning that it isn't performant for one process to wait for the other to finish.

**Pipeline Rendering** is an alternative approach that sees the main simulation logic decoupled from the render logic.
Instead of waiting on each other, the simulation logic is run on one CPU thread while the render logic is copied to a separate thread before being passed to the GPU to be rendered.
This allows the simulation to execute concurrently with the render, meaning less "idle" time for both processes.

> Bevy uses pipelined rendering, having a Main `App` that computes the simulation logic before then extracting that info into a separate Render `App`.

Reviewer: Okay this is a super nit, but `App` is a container of `SubApp`'s, one of which is the main subapp (no label, contains main `World`), and another the render subapp (`RenderApp` label, contains render `World`).

However, the Render `App` still abides by ECS principles and even has its own `World`, schedules, and systems to control how the scene is rendered.

{% callout(type="info") %}

#### GPU-Driven Vs CPU-Driven Rendering 

Determining what parts of a scene should be rendered is just as important as actually rendering the scene itself.
It's likely that some objects will obstruct or overlap others, and figuring out how those objects should be shown on screen can be an intensive process.
Like we mentioned above though, rendering has evolved to take advantage of the power that modern GPUs possess.
Instead of relying on the CPU (as was done in the past), modern rendering setups are shifting more intensive work onto the GPU itself.

This can be thought of as CPU-driven rendering versus GPU-driven rendering.
CPU-driven rendering is where the draw commands that determine how objects are rendered are created on the CPU, and only passed to the GPU to eventually be rendered.
On the other hand, GPU-driven rendering sees the GPU perform those computations itself; the CPU is only responsible for passing the data to the GPU.

> Bevy used to use CPU-driven rendering, using the CPU to choose which objects needed to be rendered and how they needed to be rendered before passing that information to the GPU.

Reviewer: Hmm, depends on what you mean, there's still CPU batching, but it's optional.

Rather than the history view, Id just say GPU is the default, but can fallback to cpu

However, this process was proving to be inefficient and was eventually changed to be GPU-driven.
Now instead of relying on the CPU to sequentially calculate each object to render, Bevy leverages GPU parallelism to perform these calculations in batches.
This helps speed up renders a lot, and reduces potential bottlenecks that might impede performance.

{% end %}

## The Rendering Process

The Render `App` follows a series of steps to carry out the rendering process.
Each step has a specific purpose in arranging and converting your games data into rendered frames.
However, it's also possible to skip certain steps if their function isn't explicitly required.

Let's look at the main four steps that occur during rendering:

### Extract

To render an image to the screen, we first have to figure out what is happening in our game!
We use the Extract step to copy information from the Main `App` and send it into the Render `App`.

> This is the first thing to occur on every frame, so reducing the amount of information that is copied can be beneficial for improving the performance of your game.

Reviewer: > This is the first thing to occur on every frame

Order isnt important, it blocks the main subapp from updating, reducing parallelism, hence you want to be quick

For example, an opaque phase describes how the objects are arranged relative to each other from the camera's perspective, and a transparency phase describes if some objects can be seen through other objects.

Remember that data that's been assigned into [`BindGroup`]s?
We use these groupings to represent the individual objects that the camera sees in each render phase (aptly called [`PhaseItem`]s).
Each `PhaseItem` is assigned a [`Draw`] function, which is the final combination of instructions for how that item will be displayed on screen.
Finally, for each `Draw` function, a draw call is created which will tell the GPU what to display on the screen.

### Render

This is where the magic finally happens!
All the queued [`Draw`] functions for every [`PhaseItem`]s are executed, incorporating the specified [`RenderPipeline`]s and [`BindGroup`]s that help us get the desired look.

Since the Render `App` abides by the ECS principles, we can leverage systems to perform the actual task of rendering.
Each system that accesses the [`RenderContext`] system parameter can be used to create the commands that are sent to the GPU to render the game.

> Using systems allows us to structure the execution of each system with a [`Schedule`], either those provided by Bevy by default ([`Core2d`] and [`Core3d`]), or by creating a custom render schedule.

Reviewer: Might be worth saying here that Bevy is camera driven - can Camera2d/Camera3d component drives a separate execution of the Core2d/Core3d schedule

Author: ~~I don't think it fits nicely in this particular section, but it's making me think that it's a better idea to put a Camera Driven Rendering page before the Render Pipelines page. I do see the benefit of mentioning it, but at the same time I feel like it distracts from the flow of the section.~~

Author: Coming back around to this, same thing as I was saying above to Alice. Don't want to get too into Camera-driven Rendering without the dedicated page to link out to. Could we leave it as is and then include a bit about it once we have the Camera Driven Rendering page?
